### PR TITLE
Require global opt-in for distributed transactions

### LINKED ISSUE #76469: Add `TransactionManager.ImplicitDistributedTransactions`
### Background and motivation

Taken from #76376:

This adds a static, global `TransactionManager.ImplicitDistributedTransactions` flag, which must be explicitly turned on in order for distributed transactions to be supported.

The reasoning here, as advocated by @ajcvickers, is that back in .NET Framework many users were unintentionally causing escalation to a distributed transaction through improper use of `TransactionScope`, etc. This flag protects users interested in using `TransactionScope` but who do not want distributed transactions; this is important as distributed transactions shouldn't be used lightly and without careful consideration (dependency on MSDTC - an external component, considerable perf impact, portability to non-Windows platforms...).

Re the flag itself, `TransactionManager` already has global settings managing transactions. Re the name `ImplicitDistributedTransactions`, there's a possibility we'd introduce cross-platform distributed transactions into `System.Transactions` (https://github.com/dotnet/runtime/issues/71769); if that happens, users will very explicitly create distributed transactions, and would not need to set this flag (hence "Implicit").

/cc @roji @ajcvickers 

### API Proposal

```C#
namespace System.Transactions;

public partial class TransactionManager
{
    [SupportedOSPlatform("windows")]
    public static bool ImplicitDistributedTransactions { get; set; }
}
```

### API Usage

```csharp
static void Main(string[] args)
{
    //...

    // Early on, the app is configured to enlist transactions into DTC:
    TransactionManager.ImplicitDistributedTransactions = true;

    //...
}
```

### PR DESCRIPTION
**Pull request by roji**
This adds a static, global TransactionManager.ImplicitDistributedTransactions flag, which must be explicitly turned on in order for distributed transactions to be supported.

The reasoning here, as advocated by @ajcvickers, is that back in .NET Framework many users were unintentionally causing escalation to a distributed transaction through improper use of TransactionScope, etc. This flag protects users interested in using TransactionScope but who do not want distributed transactions; this is important as distributed transactions shouldn't be used lightly and without careful consideration (dependency on MSDTC - an external component, considerable perf impact, portability to non-Windows platforms...).

Re the flag itself, TransactionManager already has global settings managing transactions. Re the name ImplicitDistributedTransactions, there's a possibility we'd introduce cross-platform distributed transactions into System.Transactions (#71769); if that happens, users will very explicitly create distributed transactions, and would not need to set this flag (hence "Implicit").

/cc @ajcvickers

Closes #76469

### CODE REVIEW

**Comment by stephentoub** [comment]
Is the intent here to backport this to release/7.0?

File: src/libraries/System.Transactions.Local/ref/System.Transactions.Local.cs
Comment (stephentoub) [inline src/libraries/System.Transactions.Local/ref/System.Transactions.Local.cs:194]: cc: @bartonjs, @terrajobst for the new public API

File: src/libraries/System.Transactions.Local/tests/OleTxTests.cs
Comment (stephentoub) [inline src/libraries/System.Transactions.Local/tests/OleTxTests.cs:472]: This is changing a process-wide setting.  Will that lead to race conditions based on how tests are executed?  For things where we're changing process-wide state, we generally use RemoteExecutor to isolate the changes.

Comment (roji) [reply src/libraries/System.Transactions.Local/tests/OleTxTests.cs:472]: Yeah, this is why the test that the switch works (exception unless enabled) is in the class fixture, so that it executes before everything else.

Good point though, I'll move this out using RemoteExecutor.

**Comment by roji** [comment]
> Is the intent here to backport this to release/7.0?

Yeah, this is something @ajcvickers and I just discussed, despite it being quite late. Not doing this in 7.0 would make it a breaking change later - @ajcvickers will be bringing this to tactics ASAP.

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (stephentoub) [inline src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:394]: Nit: please add XML comments (we use them to seed the docs).

**Comment by jkotas** [comment]
Can we also add this as a trimming feature flag that is off by default to allow trimming the distributed transactions code?

https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0#trimming-framework-library-features

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (carlossanlop) [reply src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:394]: And please add them in main too.

**Comment by carlossanlop** [comment]
@roji this was approved by Tactics. Pleased ping me when the unresolved of the feedback is addressed so I can merge it.

**Comment by teo-tsirpanis** [comment]
roji/runtime#3 will solve the trimming issues without adding an ILLink feature switch.

**Comment by carlossanlop** [comment]
> [roji#3](https://github.com/roji/runtime/pull/3) will solve the trimming issues without adding an ILLink feature switch.

I see that's a PR in @roji 's runtime fork. Does that mean I should wait for those commits to get merged into this PR, or do you plan on submitting that PR later as a separate 7.0 backport, making this ready to merge?

**Comment by teo-tsirpanis** [comment]
My thought is to merge my PR first and then this one, to maybe avoid going back and forth with getting servicing approval.

**Comment by roji** [comment]
@teo-tsirpanis merged your PR - thanks! Also addressed the testing and XML documentation.

Sorry for submitting this directly against release/7.0 (am used to the EF flow), I can rebase this over main or just submit a separate PR against main after this is merged.

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs
Comment (eerhardt) [inline src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs:81]: You can remove this entry now from https://github.com/dotnet/runtime/blob/992f178883f9ea79878125d64b33ef6a14a73fbd/src/libraries/System.Transactions.Local/src/ILLink/ILLink.Suppressions.LibraryBuild.xml#L4-L10

Comment (roji) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs:81]: Thanks @eerhardt, removed ILLink.Suppressions.LibraryBuild.xml.

**Comment by roji** [comment]
Rebase over latest release/7.0 (specifically #76425) and added `[SupportedOSPlatform("windows")]` on the new ImplicitDistributedTransactions flag, as discussed in #76469.

File: src/libraries/System.Transactions.Local/ref/System.Transactions.Local.cs
Comment (roji) [reply src/libraries/System.Transactions.Local/ref/System.Transactions.Local.cs]: Being discussed in #76469

File: src/libraries/System.Transactions.Local/tests/OleTxTests.cs
Comment (teo-tsirpanis) [inline src/libraries/System.Transactions.Local/tests/OleTxTests.cs:500]: Can't that be blocked on 32-bit platforms by modifying the `ConditionalFact`? I think I've seen this on other tests.

Comment (roji) [reply src/libraries/System.Transactions.Local/tests/OleTxTests.cs:500]: I admit that between IsRemoteExecutorSupported and the other "skip" stuff in this test suite, I just went for the easy thing here. Yeah, it could probably be cleaned up a bit, but it's probably not super important either.

**Comment by roji** [comment]
OK, implement the freezing behavior for ImplicitDistributedTransactions as per https://github.com/dotnet/runtime/issues/76469#issuecomment-1267337197. Also made the non-Windows version of ImplicitDistributedTransactions throw if set to true.

I think this just needs approval and we can merge.

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (jkotas) [inline src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:438]:
```suggestion
        internal static bool? s_implicitDistributedTransactions;
```
The repo naming convention for statis fields is to start with `s_`. Once you have the `s_` prefix, you do not need `Internal` to disambiguate.

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (jkotas) [inline src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:434]:
```suggestion
```
This should not be needed.

**Comment by roji** [comment]
@jkotas thanks, updated with your feedback.

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (stephentoub) [inline src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:430]: Do we need to make this setter thread-safe?  At the moment two threads racing to call set could hit issues, like double-allocating the DtcTransactionConnector.

Comment (teo-tsirpanis) [reply src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:430]: This is a class with no fields, made just to allow trimming the distributed transactions code if we don't set the flag. I don't think we should bother.

Comment (stephentoub) [reply src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:430]: That's not the only issue.  Two threads could be racing to set this, one with true, one with false.  What behavior do you expect, and how does this implementation guarantee that?

File: src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs
Comment (roji) [reply src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs:84]: You're right, this isn't needed, removing.

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (roji) [reply src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:429]: I think it was explicitly discussed  in the API design conversation that repeated setting of the same value should be allowed. I don't have any strong feelings here, but I'm also leaning towards allowing multiple sets for the same value - one could imagine an application with two independent components which attempt to set this to true, and throwing would create some needless friction there.

File: src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs
Comment (stephentoub) [inline src/libraries/System.Transactions.Local/src/System/Transactions/TransactionManager.cs:407]: Elsewhere in runtime we've used partial files for whole APIs whose implementations differ across platform.  If we use ifdefs elsewhere in System.Transactions.Local for this, it's ok.  We could also refactor it subsequently if we're trying to get this in asap.

**Review by stephentoub — APPROVED** [review]
LGTM.  Even if we don't add automated trimming tests, it'd be good to validate manually that things are being trimmed away as expected; I can't easily eyeball this to tell whether that's going to work correctly.

**Comment by roji** [comment]
I've tried to do a manual test that the code gets trimmed, and I can't see it happening. I did this by having a minimal console program with a csproj <Reference> to my locally-build System.Transactions.Local.dll with this PR's changes; running the application after publishing throws `Implicit distributed transactions have not been enabled`, which seems to prove that the DLL is getting picked up correctly.

Without setting ImplicitDistributedTransactions and publishing trimmed, System.Transaction.Local.dll is 226,304.
When setting the flag to true (and suppressing the [RequiresUnrefrerencedCode] warning it generates), System.Transaction.Local.dll is 228,864.

Inspecting the System.Transactions.Local.dll with DotPeek seems to show all the distributed transaction types - including DtcTransactionConnector - although these should be trimmed.

/cc @teo-tsirpanis @eerhardt 

<details>
<summary>Console program</summary>

```c#
// #pragma warning disable IL2026
// TransactionManager.ImplicitDistributedTransactions = true;
// #pragma warning restore IL2026

Console.WriteLine(typeof(TransactionScope).Assembly.Location);

await using var conn1 = new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false");
await using var conn2 = new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false");

using (var scope = new TransactionScope())
{
    conn1.Open();
    conn2.Open();
}
```

csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net7.0-windows</TargetFramework>
        <LangVersion>latest</LangVersion>
        <Nullable>enable</Nullable>
        <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
        <PublishTrimmed>true</PublishTrimmed>
        <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Microsoft.Data.SqlClient" Version="3.0.1" />
    </ItemGroup>

    <ItemGroup>
        <Reference Include="C:\projects\runtime\artifacts\bin\System.Transactions.Local\Debug\net7.0-windows\System.Transactions.Local.dll" />
    </ItemGroup>

</Project>
```

</details>

**Comment by teo-tsirpanis** [comment]
You should check which type references `DtcTransactionConnector`. ILSpy has an "Analyze" option that will help you quickly find out.

**Comment by stephentoub** [comment]
@teo-tsirpanis, when you prototyped it, did you validate everything was getting trimmed away?

**Comment by teo-tsirpanis** [comment]
Unfortunately not. 😢 I thought it would be OK since trimming warnings were unsuppressed and did not show up.

**Comment by stephentoub** [comment]
Yeah, I don't see how this construction would enable trimming.

**Comment by teo-tsirpanis** [comment]
I explained how (I thought) it works in https://github.com/roji/runtime/pull/3#issuecomment-1263963781. Was any of my assumptions wrong?

**Comment by teo-tsirpanis** [comment]
I wrote a tiny program that follows the same logic and it works.

```csharp
using ClassLibrary1;

namespace TrimTest
{
    internal static class Program
    {
        static void Main()
        {
            Console.WriteLine("Hello, World!");
            Call(null);
        }

        static void Call(I? i) => i?.F();
    }
}
```

__In another library__

```csharp
namespace ClassLibrary1
{
    public interface I { void F(); }

    public class C : I { public void F() { } }
}
```

`C` did not get created and got trimmed as expected. Similarly if `DtcTransactionConnector` does not get created (if we don't set the switch to true), it would have not been created and should have been trimmed. Unless something else keeps it. 🤔

@roji if you can send the published trimmed app I will take a look.

**Comment by roji** [comment]
@teo-tsirpanis sent you an email offline, thanks!

Also squashed and rebased over latest release/7.0.

**Comment by teo-tsirpanis** [comment]
Thanks for the email @roji. I inspected the trimmed assemblies and `DtcTransactionConnector` is not there. Nor is `TransactionManager.ImplicitDistributedTransactions`. Perhaps DotPeek decided to show you a different assembly (I almost never use it, ILSpy is much better).

The trim-unsafe code, the P/Invoke that does the COM interop is removed. Removing more code will need additional refactorings.

image

**Comment by roji** [comment]
@teo-tsirpanis thanks, yeah - I mistakenly looked at ITransactionConnector instead of DtcTransactionConnector, I can indeed see the latter getting trimmed (and thanks for the ILSpy tip, it's indeed very useful also for understanding who's referencing a type, keeping it from getting trimmed).

However, as the actual assembly sizes show (2.5k diff), this change alone doesn't allow any significant trimming as there are other references. In order to get real benefits here, it seems that all the trimmable code would have to be refactored into ITransactionConnector. As we're snapping for 7.0 GA on Monday, I've opened #76791 to track doing this for 8.0.

@carlossanlop this is now ready to merge as-is.

**Comment by carlossanlop** [comment]
CI re-run passed, all green. Approved and signed-off. Ready to merge. :shipit:

**Comment by roji** [comment]
@carlossanlop the PR adds XML docs to the new API (ImplicitDistributedTransactions), but do I need to update the XMLs manually in https://github.com/dotnet/dotnet-api-docs?

**Comment by carlossanlop** [comment]
dotnet-api-docs is the source of truth, so yes, please add it there too.

**Comment by carlossanlop** [comment]
You can port the triple slash docs automatically to dotnet-api-docs using the `PortToDocs.exe` tool. You can get it from the https://github.com/dotnet/api-docs-sync repo, and you can install it using the `install-as-tool.ps1` script located in the root of the repo.

You would use a command like this:

```
PortToDocs.exe -Docs %RepoRoot%\dotnet-api-docs\xml -IntelliSense %RepoRoot%\runtime\artifacts\bin -IncludedAssemblies "YourAssemblyHere" -IncludedNamespaces "YourNamespaceHere" -Save true
```

**Pull request merged**
